### PR TITLE
Setup GitHub Actions CI and SNAPSHOT uploading

### PR DESCRIPTION
**Pull request by lazaroclapp**
This change sets up GitHub Actions as a parallel build CI to Travis (for now). GitHub Actions should build NullAway on MacOS X and Linux using JDK 8, and on Linux using JDK 11, mirroring our existing Travis config. 

Additionally, this migrates snapshot uploading and coverage tracking to GitHub Actions, disabling the equivalent Travis CI events as they will likely conflict. This might break snapshot uploading after landing, since it's not possible to easily test that from the PR. If it does, follow up PRs might be needed.

See #408

### CODE REVIEW

**Comment by lazaroclapp** [comment]
Do not merge yet. WIP. Testing.

**Comment by lazaroclapp** [comment]
@msridhar No idea if Snapshot publishing works yet, but GitHub action CI builds themselves look sane :)

File: .github/workflows/continuous-integration.yml
Comment (msridhar) [inline .github/workflows/continuous-integration.yml:74]: For WALA I had to do this to get env vars to show up as Gradle properties in the build:

```
          ORG_GRADLE_PROJECT_SONATYPE_NEXUS_USERNAME: ${{ secrets.SONATYPE_NEXUS_USERNAME }}
          ORG_GRADLE_PROJECT_SONATYPE_NEXUS_PASSWORD: ${{ secrets.SONATYPE_NEXUS_PASSWORD }}
```
But the WALA Gradle scripts are different from NullAway's so maybe this is correct here.

**Review by msridhar — APPROVED** [review]
Overall looks great!  Just a couple non-blocking comments

**Comment by msridhar** [comment]
BTW, I was a little surprised this "just worked" without doing anything explicit to set up the Android SDK in the builds.  Looks like Android [comes pre-installed](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md) on the GH Actions runners; nice!

**Comment by lazaroclapp** [comment]
> BTW, I was a little surprised this "just worked" without doing anything explicit to set up the Android SDK in the builds. Looks like Android [comes pre-installed](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md) on the GH Actions runners; nice!

Yup. I had to read into that. Was similarly confused for a while, could have probably saved you the time in searching, though 😅

File: .github/workflows/continuous-integration.yml
Comment (lazaroclapp) [reply .github/workflows/continuous-integration.yml:44]: Ah, let me look into that a bit. Might be ok, but we are not in a rush.

File: .github/workflows/continuous-integration.yml
Comment (lazaroclapp) [reply .github/workflows/continuous-integration.yml:74]: Our scripts look for `SONATYPE_NEXUS_USERNAME`, etc, in env vars. So I suspect this will work, but let me test it.

Comment (lazaroclapp) [reply .github/workflows/continuous-integration.yml]: Actually, you were right, `ORG_GRADLE_PROJECT_SONATYPE_NEXUS_USERNAME` works for some reason and `SONATYPE_NEXUS_USERNAME` doesn't. I could check out why, but since it seems to be working now, I'll just update the branch and land.

Comment (msridhar) [reply .github/workflows/continuous-integration.yml]: It's because the Gradle scripts are looking for a project property `SONATYPE_NEXUS_USERNAME`, etc., and you need the `ORG_GRADLE_PROJECT_` prefix to have Gradle read the environment variable as a project property: https://docs.gradle.org/current/userguide/build_environment.html#sec:project_properties

**Comment by msridhar** [comment]
BTW does each PR creation really imply a push?  What if it's a PR from someone without push access to this repo?

**Comment by lazaroclapp** [comment]
> BTW does each PR creation really imply a push? What if it's a PR from someone without push access to this repo?

Not sure. I guess the only way to test it is to try to create a PR from a fork or some other repo. The problem with `pull_request` (as configured in WALA) is that it will only trigger at PR creation and it will duplicate what we see from just push (so you get twice as many jobs in the PR). There is a way to also trigger when the PR changes, but again that will cause these jobs to show twice... I'd want to test if nothing shows right now for a PR from a different repo before enabling that much redundancy.

Might experiment with that tomorrow (also, there seems to be still a step missing in uploading coverage results :) )

File: .github/workflows/continuous-integration.yml
Comment (joschi) [reply .github/workflows/continuous-integration.yml:44]: Since the author of [kt3k/coveralls-gradle-plugin](https://github.com/kt3k/coveralls-gradle-plugin) is looking for a new maintainer (see https://github.com/kt3k/coveralls-gradle-plugin/issues/107) and [nbaztec/coveralls-jacoco-gradle-plugin](https://github.com/nbaztec/coveralls-jacoco-gradle-plugin) supports GitHub Actions out-of-the-box, it might be worth migrating to the latter.

https://github.com/nbaztec/coveralls-jacoco-gradle-plugin#github-actions

File: .github/workflows/continuous-integration.yml
Comment (joschi) [inline .github/workflows/continuous-integration.yml:67]: Using [eskatos/gradle-command-action](https://github.com/eskatos/gradle-command-action) might reduce the boilerplate necessary to set up proper caching for Gradle builds.

https://github.com/eskatos is working for Gradle, so it's not some random 3rd party GitHub action.

File: jar-infer/nullaway-integration-test/build.gradle
Comment (lazaroclapp) [inline jar-infer/nullaway-integration-test/build.gradle:20]: Coverage plugin has [multi-project support](https://github.com/nbaztec/coveralls-jacoco-gradle-plugin#multi-project-support---android). But:
a) It's a bit convoluted
b) We have never had proper coverage support for JarInfer before (this bit above is long broken)
c) We mostly care about coverage for core NullAway

That said, it's worth having JarInfer coverage in the future as well, just not high priority.

File: .github/workflows/continuous-integration.yml
Comment (lazaroclapp) [reply .github/workflows/continuous-integration.yml:67]: Makes sense to me. Thanks! :)

**Comment by lazaroclapp** [comment]
Coveralls plug-in was sometimes failing with a timeout. Not sure if this will keep happening after merging or if the last change I made truly fixed it (ran it twice, successfully), but I'd say let's keep an eye on it and fix as needed.

Also, not sure if the GitHub Actions CI will trigger for third-party PRs. Since we are running both GitHub Actions *and* Travis CI for the time being, I suggest we just keep an eye out for anything that seems to be skipping GitHub Actions.

Will merge unless there are concerns from @msridhar

**Comment by msridhar** [comment]
LGTM; I'm fine with landing.  My feeling is that CI won't run for 3rd-party PRs right now, but we can easily test that once this lands

**Pull request merged**
